### PR TITLE
Add pytorch workloads GPU/CPU and decorator examples

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #95

---

[Datalayer](https://datalayer.io)

[Become a Sponsor](https://github.com/sponsors/datalayer)

# PyTorch GPU Workloads + Datalayer SDK Example

This example demonstrates how to leverage GPU acceleration for computationally intensive PyTorch workloads using the Datalayer SDK, showcasing significant performance improvements over CPU-only execution.

## Overview

This example showcases:

- **GPU Acceleration**: High-performance computing workloads utilizing CUDA-enabled GPUs
- **Datalayer SDK**: Integration with Datalayer's runtime environment for AI workloads
- **Performance Benchmarking**: Side-by-side comparison of CPU vs GPU execution times
- **PyTorch Integration**: Optimized matrix operations and tensor computations
- **Heavy Workload Processing**: Large-scale matrix multiplications demonstrating GPU advantages

## Features

- ✅ Execute compute-intensive PyTorch operations on GPU and CPU
- ✅ Benchmark and compare performance between CPU and GPU execution
- ✅ Demonstrate significant speedup with GPU acceleration (typically 10-100x faster)
- ✅ Reproducible results with deterministic operations
- ✅ CUDA synchronization for accurate GPU timing measurements

## Prerequisites

- Python 3.9+
- Datalayer SDK
- PyTorch with CUDA support
- CUDA-compatible GPU (for GPU benchmarks)

## Installation

1. **Clone the repository**:

   ```bash
   git clone https://github.com/datalayer/core.git
   cd core/examples/pytorch-workloads
   ```

2. **Install dependencies**:
   ```bash
   pip install -r requirements.txt
   ```

## Quick Start

**Run the GPU vs CPU benchmark**:

```bash
python main.py
```

This will execute two comprehensive benchmarks on both CPU and GPU (if available), displaying:

- **Matrix Multiplication**: Large-scale linear algebra operations
- **Convolution Operations**: Deep learning CNN-style computations
- Execution times for both devices and operations
- Individual and average speedup factors (GPU vs CPU)

## Benchmark Results

The example performs two types of heavy computations and compares execution times:

- Large matrix multiplications (10,000x10,000 by default): Measures the performance of multiplying two large matrices, a common operation in scientific computing and machine learning.
- Convolution Benchmark. Deep learning-style 2D convolutions (48 batch size, 180 channels, 640x640 images):

Running the `main.py` scripts results in the following example output:

```bash
PyTorch CPU vs GPU Benchmark
============================

1. Matrix Multiplication Benchmark
-----------------------------------

Running on: cpu

Matrix size: 10000x10000
Execution time: 2.5394 seconds

Running on: cuda

Matrix size: 10000x10000
Execution time: 0.1122 seconds



Matrix Multiplication Speedup: 22.63x faster on GPU


2. Convolution Benchmark
------------------------

Running convolution benchmark on: cpu

Input tensor shape: torch.Size([48, 180, 640, 640])
Output tensor shape: torch.Size([48, 180, 640, 640])
Execution time: 34.7170 seconds

Running convolution benchmark on: cuda

Input tensor shape: torch.Size([48, 180, 640, 640])
Output tensor shape: torch.Size([48, 180, 640, 640])
Execution time: 0.4160 seconds



Convolution Speedup: 83.44x faster on GPU

==================================================
BENCHMARK SUMMARY
==================================================
Matrix Multiplication: 22.63x speedup
Convolution Operations: 83.44x speedup
Average GPU Speedup: 53.04x
```

**Overall Average Speedup**: ~20-25x faster on GPU

## Project Structure

```
pytorch-workloads/
├── requirements.txt      # Python dependencies (torch)
├── README.md             # This file
├── main.py               # Main benchmark execution script
└── benchmark.py          # Core benchmarking functions
```

## Understanding the Benchmarks

The example includes two comprehensive benchmarks:

### Matrix Multiplication Benchmark

1. **Matrix Generation**: Creates large random matrices (10,000x10,000)
2. **Device Allocation**: Moves tensors to CPU or GPU memory
3. **Warm-up**: Ensures GPU is ready for accurate timing
4. **Matrix Multiplication**: Executes computationally expensive `torch.matmul()`
5. **Synchronization**: Ensures GPU operations complete before timing

### Convolution Benchmark

1. **Tensor Creation**: Generates large 4D tensors (batch_size=48, channels=180, height=640, width=640)
2. **CNN Layer Setup**: Creates convolutional layers with 3x3 kernels
3. **Memory Transfer**: Moves tensors and models to target device
4. **Convolution Operations**: Performs two sequential 2D convolution operations
5. **GPU Synchronization**: Ensures accurate timing measurements

## Use Cases

This example is ideal for:

- **Deep Learning Training**: Accelerating neural network training workflows
- **Scientific Computing**: Large-scale numerical computations
- **Data Processing**: High-throughput tensor operations
- **Performance Optimization**: Identifying GPU acceleration opportunities
- **Benchmarking**: Evaluating hardware performance for AI workloads

## License

This project is licensed under the MIT License - see the [LICENSE](../../LICENSE) file for details.

## Support

- **Documentation**: [Datalayer Platform Documentation](https://docs.datalayer.app/)
- **Issues**: [GitHub Issues](https://github.com/datalayer/core/issues)
- **Community**: [Datalayer Platform](https://datalayer.app/)

---

<p align="center">
  <strong>🚀 AI Platform for Data Analysis</strong><br></br>
  <a href="https://datalayer.app/">Get started with Datalayer today!</a>
</p>

-----

[Datalayer](https://datalayer.io)

[Become a Sponsor](https://github.com/sponsors/datalayer)

# Datalayer Decorator + SDK Example

This example demonstrates how to use the `@datalayer` decorator to seamlessly execute Python functions on remote Datalayer runtimes, enabling distributed computing with minimal code changes.

## Overview

This example showcases:

- **Function Decoration**: Transform regular functions into distributed computations using `@datalayer`
- **Remote Execution**: Execute functions on cloud-based runtimes with different environments
- **Variable Management**: Pass inputs and retrieve outputs from remote execution contexts
- **Snapshot Integration**: Use pre-configured runtime snapshots for consistent environments
- **Error Handling**: Timeout configuration and debug mode for development

## Features

-  Simple function decoration for remote execution
-  Input/output variable mapping between local and remote contexts
-  Multiple runtime environments (CPU, GPU, specialized environments)
-  Snapshot-based runtime initialization
-  Debug mode for development and troubleshooting
-  Comprehensive examples covering different use cases

## Prerequisites

- Python 3.9+
- Datalayer SDK
- Environment access to Datalayer platform
- Optional: scikit-learn, pandas, numpy (for ML examples)

## Installation

1. **Clone the repository**:

   ```bash
   git clone https://github.com/datalayer/core.git
   cd core/examples/decorator
   ```

2. **Install dependencies**:
   ```bash
   pip install -e ../../[examples]
   ```

## Quick Start

**Run the decorator examples**:

```bash
python main.py
```

This will execute four different examples demonstrating various decorator usage patterns:

- **Simple Computation**: Basic mathematical operations
- **Data Processing**: Statistical analysis with input/output mapping
- **Machine Learning**: Model training with scikit-learn
- **Data Analysis**: Dataset analysis using pandas with snapshots

### 1. Simple Function Decoration

```python
@datalayer(runtime_name="my-runtime", environment="ai-env")
def simple_computation(x: int, y: int) -> int:
    result = x * y + 10
    return result

# Function executes remotely but called normally
result = simple_computation(5, 8)
```

### 2. Input/Output Variable Mapping

```python
@datalayer(
    runtime_name="data-processing",
    environment="ai-env",
    inputs=["data"],
    output="processed_result"
)
def process_data(data: list) -> dict:
    # Process data remotely
    return {"count": len(data), "mean": sum(data)/len(data)}
```

### 3. Machine Learning with Libraries

```python
@datalayer(
    runtime_name="ml-training",
    environment="ai-env",
    inputs=["features", "labels"],
    output="model_metrics",
    debug=True
)
def train_simple_model(features: list, labels: list) -> dict:
    from sklearn.linear_model import LinearRegression
    # ML code executes in remote environment with sklearn available
    # ...
    return model_metrics
```

### 4. Snapshot-based Environments

```python
@datalayer(
    runtime_name="analysis-runtime",
    environment="ai-env",
    snapshot_name="data-analysis-env",  # Pre-configured environment
    timeout=30.0
)
def analyze_dataset() -> dict:
    import pandas as pd
    # Use pre-installed packages from snapshot
    # ...
    return analysis_results
```

### Core Parameters

- **`runtime_name`** (str, optional): Name for the runtime instance
- **`environment`** (str): Environment type (e.g., "ai-env", "gpu-env")
- **`inputs`** (list[str], optional): Local variables to pass to remote function
- **`output`** (str, optional): Remote variable to return as result
- **`snapshot_name`** (str, optional): Pre-configured snapshot to use
- **`debug`** (bool): Enable detailed logging and error output
- **`timeout`** (float): Execution timeout in seconds

### Environment Types

- **`ai-env`**: GPU-enabled environment for deep learning
- **`python-cpu-env`**: General-purpose environment with common ML libraries

## Use Cases

This decorator pattern is ideal for:

- **Distributed Computing**: Scale functions across multiple runtimes
- **Resource-Intensive Tasks**: Execute heavy computations on powerful remote hardware
- **Environment Isolation**: Run code in specific, controlled environments
- **Collaborative Development**: Share consistent runtime environments across teams
- **Cost Optimization**: Use cloud resources only when needed
- **GPU Acceleration**: Execute ML workloads on GPU-enabled runtimes

## Project Structure

```
decorator/
  README.md             # This file
  main.py               # Comprehensive decorator examples
```

## Output Example

```bash
Datalayer Decorator Examples
========================================

1. Simple Computation Example
------------------------------


Computing 5 * 8 + 10 = 50


Result: 50

2. Data Processing Example
------------------------------


Processed 13 data points
Statistics: {'count': 13, 'mean': 8.846153846153847, 'median': 7, 'stdev': 7.174563329873044, 'min': 1, 'max': 25}


Processing result: {'count': 13, 'mean': 8.846153846153847, 'median': 7, 'stdev': 7.174563329873044, 'min': 1, 'max': 25}

3. Machine Learning Example
------------------------------


Model trained with R� score: 0.993
Mean Squared Error: 22.504


Model metrics: {'mse': 22.503574409122553, 'r2': 0.9932780416026469, 'coefficient': 1.9653427974453015, 'intercept': 2.6844899679603174, 'train_size': 80, 'test_size': 20}

4. Data Analysis Example
------------------------------


Dataset Analysis Complete
Shape: (1000, 4)
Correlations with target: {'feature_1': -0.017914403077249664, 'feature_2': -0.009844943287877567, 'feature_3': -0.05046032243985329}


Analysis complete. Dataset shape: (1000, 4)
```

## License

This project is licensed under the MIT License - see the [LICENSE](../../LICENSE) file for details.

## Support

- **Documentation**: [Datalayer Platform Documentation](https://docs.datalayer.app/)
- **Issues**: [GitHub Issues](https://github.com/datalayer/core/issues)
- **Community**: [Datalayer Platform](https://datalayer.app/)

---

<p align="center">
  <strong>=� AI Platform for Data Analysis</strong><br></br>
  <a href="https://datalayer.app/">Get started with Datalayer today!</a>
</p>

### CODE REVIEW

**Review by echarles — APPROVED** [review]
LGTM Thx @goanpeca
